### PR TITLE
Descriptor struct overhaul

### PR DESCRIPTION
**Pull request by sanket1729**
As discussed, this refactors descriptor module. We have a `DescriptorTrait` which we implement for all the basic descriptors. The internal descriptor structs are also more organized into separate files and clean implementations with code reuse where things make sense. 

Only the following should change for library user:
1) Instead of creating a Descriptor from an enum, the user now has to call the `new` functions for creating descriptors. 
2) Users would have to import the trait `DescriptorTrait` to use the functions.

### CODE REVIEW

File: src/descriptor/mod.rs
Comment (darosior) [inline src/descriptor/mod.rs:102]: When would we ever return None ? Especially since below we don't return an Option for the scriptPubKey.

**Review by darosior — COMMENTED** [review]
Concept ACK, even if that means YA big refactor downstream :( 

Quickly skimmed through it and left a few comment. Will review it more in depth soon :tm:. Overall it really is clean and neat

File: src/descriptor/mod.rs
Comment (sanket1729) [reply src/descriptor/mod.rs:102]: In the case of Bare Descriptor, we return `None`. But as another PR, after this is merged I plan to change all option returning APIs to return Results.

**Comment by dr-orlovsky** [comment]
Looks nice to me, as a concept! (and yes, yet another big refactoring, but it worth it)

Have you thought about providing a simple enumeration with descriptor types? Some software may present with options which descriptors are supported, and having an enumeration w/o embedded descriptor data may be useful. Also a `DesctiptorTrait::descriptor_type` method returning that contentless type may be useful

**Comment by sanket1729** [comment]
@dr-orlovsky, I like the idea of the descriptor type enum without the content. I think we should do something like this.

**Comment by dr-orlovsky** [comment]
@sanket1729 let me work on the PR (I will make it against the master, but it will be easy to rebase on top of it or merge it into your PR). I had a lot of requirements for the functionality such descriptor types and I have already did a few attempts in different projects like [Bitcoin Pro](https://github.com/pandoracore/bitcoin-pro)

File: src/descriptor/mod.rs
Comment (apoelstra) [inline src/descriptor/mod.rs:79]: We can probably drop most (all?) of these requirements.

File: src/descriptor/mod.rs
Comment (apoelstra) [inline src/descriptor/mod.rs:145]: Should name this differently, say `explicit_script` or something ... and it should also be a `Result` since Taproot descriptors won't have anything meaningful to put here.

File: src/descriptor/mod.rs
Comment (apoelstra) [inline src/descriptor/mod.rs:215]: s/must/should/

File: src/descriptor/mod.rs
Comment (sanket1729) [reply src/descriptor/mod.rs]: I think we should leave this to other changes in Taproot descriptors. We may understand more things as we understand and implement them.

File: src/descriptor/mod.rs
Comment (sanket1729) [reply src/descriptor/mod.rs:79]: I removed the bounds. But unfortunately, removing them did not help that much with code ergonomics.

**Review by apoelstra — APPROVED** [review]
ack f8a9478349ad660fdae1c7d4931daeb6deb0250e

**Pull request merged**
